src/table_builder.cc

```diff
@@ -52,18 +53,36 @@ void TitanTableBuilder::Add(const Slice& key, const Slice& value) {
       base_builder_->Add(key, value);
     }
   } else if (ikey.type == kTypeValue &&
-             value.size() >= cf_options_.min_blob_size &&
              cf_options_.blob_run_mode == TitanBlobRunMode::kNormal) {
-    // we write to blob file and insert index
-    std::string index_value;
-    AddBlob(ikey.user_key, value, &index_value);
-    UpdateIOBytes(prev_bytes_read, prev_bytes_written, &io_bytes_read_,
-                  &io_bytes_written_);
-    if (ok()) {
-      ikey.type = kTypeBlobIndex;
-      std::string index_key;
-      AppendInternalKey(&index_key, ikey);
-      base_builder_->Add(index_key, index_value);
+    bool is_small_kv = value.size() < cf_options_.min_blob_size;
+    if (is_small_kv &&
+        (!blob_builder_ || blob_builder_->GetBuilderState() ==
+                               BlobFileBuilder::BuilderState::kUnbuffered)) {
+      // We can append this into SST safely, without disorder issue.
+      base_builder_->Add(key, value);
+      return;
+    } else if (is_small_kv) {
+      // We have to let builder to cache this KV pair, and it will be returned
+      // when state changed
+      BlobRecord record;
+      record.key = ikey.user_key;
+      record.value = value;
+      std::unique_ptr<BlobFileBuilder::BlobRecordContext> ctx(
+          new BlobFileBuilder::BlobRecordContext);
+      AppendInternalKey(&ctx->key, ikey);
+      blob_builder_->CacheSmallKV(record, std::move(ctx));
+    } else {
+      // We write to blob file and insert index
+      BlobRecord record;
+      record.key = ikey.user_key;
+      record.value = value;
+      BlobFileBuilder::OutContexts contexts;
+      AddBlob(record, ikey, &contexts);
+      AddToBaseTable(contexts);
+      UpdateIOBytes(prev_bytes_read, prev_bytes_written, &io_bytes_read_,
+                    &io_bytes_written_);
+
+      if (ok()) return;
     }
   } else if (ikey.type == kTypeBlobIndex && cf_options_.level_merge &&
```

Comment:
For all `base_builder_->Add`, you should check the state of the blob builder. It still suffers from the disorder issue in the case that it is `kTypeBlobIndex` but doesn't hit this if branch.

Reply:
Please update the disorder test case too